scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/AvroBucketMetadata.java

```diff
@@ -35,50 +36,67 @@
 /**
  * {@link org.apache.beam.sdk.extensions.smb.BucketMetadata} for Avro {@link GenericRecord} records.
  */
-public class AvroBucketMetadata<K, V extends GenericRecord> extends BucketMetadata<K, V> {
+public class AvroBucketMetadata<K1, K2, V extends GenericRecord> extends BucketMetadata<K1, K2, V> {
 
   @JsonProperty private final String keyField;
+  @JsonProperty private final String keyFieldSecondary;
 
   @JsonIgnore private final String[] keyPath;
+  @JsonIgnore private final String[] keyPathSecondary;
 
   public AvroBucketMetadata(
       int numBuckets,
       int numShards,
-      Class<K> keyClass,
+      Class<K1> keyClassPrimary,
+      Class<K2> keyClassSecondary,
       BucketMetadata.HashType hashType,
       String keyField,
+      String keyFieldSecondary,
```

Comment:
Wouldn't it make more sense to group the class and the field together?
Field ordering is already not really respected with the base class since `keyField` appear before the `filenamePrefix`.

Also, having primary & secondary as prefix would look more natural on the reading and favorises grouping

```suggestion
      Class<K1> primaryKeyClass,
      String primaryKeyField,
      Class<K2> secondaryKeyClass,
      String secondaryKeyField,
      BucketMetadata.HashType hashType,
```

Reply:
This is all internal so should basically never be constructed by a user; I've just maintained the existing order.

I evaluated using `primaryKeyClass`/`secondaryKeyClass` but the diff was harder to follow since the primary/secondary info is a secondary concern to whether you're operating on the class vs field etc.

Reply:
> the primary/secondary info is a secondary concern whether you're operating on the class vs field.

IMHO we'll actually operate more often at primary/secondary level rather that key/field

Also in term of validity, in the avro and json wold, the keys and fields are strongly coupled and can't live without the other as seen in the assertion bellow, they both belong to the primary/secondary umbrella (which could actually be a key extractor class). For other storage, we could imagine that other property can come into play.

Agreed for the diff. This is already big enough, we should'nt make it more complex

Reply:
yeah this is a tricky one. I understand not wanting to make it more complex, but I think Michel's suggested ordering is more in line with standard Scala/Java APIs - for example   [Map](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/immutable/Map.scala#L289) [constructors](https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/util/Map.html#of(K,V,K,V,K,V)) are ordered like `k1, v1, k2, v2, ...` not `k1, k2, v1, v2`.

Reply:
Done